interactions/utils/abc/base_iterators.py

```diff
@@ -0,0 +1,97 @@
+from abc import ABCMeta, abstractmethod
+from math import inf
+from typing import TYPE_CHECKING, Callable, List, Optional, TypeVar, Union
+
+from ..missing import MISSING
+
+_T = TypeVar("_T")
+_O = TypeVar("_O")
+
+__all__ = ("BaseAsyncIterator", "BaseIterator")
+
+if TYPE_CHECKING:
+    from ...api.http.client import HTTPClient
+    from ...api.models.misc import Snowflake
+
+
+class BaseAsyncIterator(metaclass=ABCMeta):
+    """A base class for async iterators."""
+
+    # I don't want to make it subclass the BaseIterator since it forces implementation of __next__ and __iter__
+
+    @abstractmethod
+    def __init__(
+        self,
+        obj: Union[int, str, "Snowflake", _T],
+        _client: Optional["HTTPClient"] = None,
+        maximum: Optional[int] = inf,
+        start_at: Optional[Union[int, str, "Snowflake", _O]] = MISSING,
+        check: Optional[Callable[[_O], bool]] = None,
+    ):
+
+        self.object_id = int(obj) if not hasattr(obj, "id") else int(obj.id)
+        self.maximum = maximum
+        self.check = check
+        self._client = _client
+        self.object_count: int = 0
+        self.start_at = (
+            None
+            if start_at is MISSING
+            else int(start_at)
+            if not hasattr(start_at, "id")
+            else int(start_at.id)
+        )
+        self.__stop: bool = False
+        self.objects: Optional[List[_O]] = None
+
+    @abstractmethod
+    async def get_first_objects(self) -> None:
```

Comment:
Not every iterator will have a difference between getting the first object and getting the subsequent objects, or even need another method to do so

Reply:
You can just leave one of the methods with an ellipsis defined in your class

Reply:
You can't leave the method unoverriden without errors. Adding the function with no code can easily lead to confusion